File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/AgentEntitlementsStore.swift
import Foundation
import Observation

/// `AgentEntitlementsStore` ports `wagerproof-mobile/hooks/useAgentEntitlements.ts`
/// — the freemium/pro gating layer for agent creation, viewing picks, and
/// leaderboard ranks.
///
/// Backed by `ProAccessStore` so changes to the user's RC entitlement
/// propagate automatically.
///
/// Limits mirror the RN constants:
///   - FREE: 1 active agent
///   - PRO: 10 active / 30 total
///   - Admin: unlimited
@Observable
@MainActor
public final class AgentEntitlementsStore {
    public static let freeAgentLimit = 1
    public static let proMaxActiveAgents = 10
    public static let proMaxTotalAgents = 30
    public static let freeLeaderboardMinRank = 6
    public static let freeLeaderboardMaxRank = 10
    /// Hard cap on how many agents may be *active* at once, for any tier — the
    /// office has 8 desks, so at most 8 agents run concurrently. Users may still
    /// create up to the total-agent limit; the rest sit inactive on the bench.
    public static let maxConcurrentActiveAgents = 8

    private let proAccess: ProAccessStore

    public init(proAccess: ProAccessStore) {
        self.proAccess = proAccess
    }

    public var isPro: Bool { proAccess.isPro }
    public var isAdmin: Bool { proAccess.isAdmin }
    public var isLoading: Bool { proAccess.isLoading }

    public var canViewAgentPicks: Bool { isPro || isAdmin }
    public var canCreatePublicAgent: Bool { isPro || isAdmin }
    public var canUseAutopilot: Bool { isPro || isAdmin }

    public var maxActiveAgents: Int? {
        if isAdmin { return nil }
        return isPro ? Self.proMaxActiveAgents : Self.freeAgentLimit
    }

    public var maxTotalAgents: Int? {
        if isAdmin { return nil }
        return isPro ? Self.proMaxTotalAgents : Self.freeAgentLimit
    }

    /// Mirrors RN `canCreateAnotherAgent`. Pro users gate on TOTAL count,
    /// free users gate on ACTIVE count.
    public func canCreateAnotherAgent(activeCount: Int, totalCount: Int) -> Bool {
        if isAdmin { return true }
        if isPro { return totalCount < Self.proMaxTotalAgents }
        return activeCount < Self.freeAgentLimit
    }

    /// Mirrors RN `canViewLeaderboardRank`. Free users can preview ranks 6–10
    /// (the "you could be here" tease); Pro/admin see all.
    public func canViewLeaderboardRank(_ rank: Int) -> Bool {
        if isPro || isAdmin { return true }
        return rank >= Self.freeLeaderboardMinRank && rank <= Self.freeLeaderboardMaxRank
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/AgentEntitlementsStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable

/**
 * Port of iOS `AgentEntitlementsStore.swift` (doc §4.3).
 *
 * Freemium/Pro gating for agent creation, pick visibility, and leaderboard
 * ranks. Backed by [ProAccessStore] so RC entitlement changes propagate.
 *
 * Limits: FREE = 1 active agent; PRO = 10 active / 30 total; Admin = unlimited.
 */
@Stable
class AgentEntitlementsStore(
    private val proAccess: ProAccessStore,
) {
    companion object {
        const val FREE_AGENT_LIMIT = 1
        const val PRO_MAX_ACTIVE_AGENTS = 10
        const val PRO_MAX_TOTAL_AGENTS = 30
        const val FREE_LEADERBOARD_MIN_RANK = 6
        const val FREE_LEADERBOARD_MAX_RANK = 10

        /** Hard cap on concurrently-active agents for any tier ("8 desks"). */
        const val MAX_CONCURRENT_ACTIVE_AGENTS = 8
    }

    val isPro: Boolean get() = proAccess.isPro
    val isAdmin: Boolean get() = proAccess.isAdmin
    val isLoading: Boolean get() = proAccess.isLoading

    val canViewAgentPicks: Boolean get() = isPro || isAdmin
    val canCreatePublicAgent: Boolean get() = isPro || isAdmin
    val canUseAutopilot: Boolean get() = isPro || isAdmin

    val maxActiveAgents: Int?
        get() = if (isAdmin) null else if (isPro) PRO_MAX_ACTIVE_AGENTS else FREE_AGENT_LIMIT

    val maxTotalAgents: Int?
        get() = if (isAdmin) null else if (isPro) PRO_MAX_TOTAL_AGENTS else FREE_AGENT_LIMIT

    /** Pro users gate on TOTAL count, free users gate on ACTIVE count. */
    fun canCreateAnotherAgent(activeCount: Int, totalCount: Int): Boolean {
        if (isAdmin) return true
        if (isPro) return totalCount < PRO_MAX_TOTAL_AGENTS
        return activeCount < FREE_AGENT_LIMIT
    }

    /** Free users can preview ranks 6–10 (the "you could be here" tease); Pro/admin see all. */
    fun canViewLeaderboardRank(rank: Int): Boolean {
        if (isPro || isAdmin) return true
        return rank in FREE_LEADERBOARD_MIN_RANK..FREE_LEADERBOARD_MAX_RANK
    }
}
